pageserver/src/walreceiver.rs

```diff
@@ -64,375 +80,1133 @@ thread_local! {
+    WAL_RECEIVER_ENTRIES
+        .read()
+        .await
+        .get(&ZTenantTimelineId::new(tenant_id, timeline_id))
+        .cloned()
 }
 
-// Launch a new WAL receiver, or tell one that's running about change in connection string
-pub fn launch_wal_receiver(
+/// Sets up the main WAL receiver thread that manages the rest of the subtasks inside of it, per timeline.
+/// See comments in [`wal_receiver_main_thread_loop_step`] for more details on per timeline activities.
+pub fn init_wal_receiver_main_thread(
     conf: &'static PageServerConf,
-    tenantid: ZTenantId,
-    timelineid: ZTimelineId,
-    wal_producer_connstr: &str,
-) -> Result<()> {
-    let mut receivers = WAL_RECEIVERS.lock().unwrap();
-
-    match receivers.get_mut(&(tenantid, timelineid)) {
-        Some(receiver) => {
-            debug!("wal receiver already running, updating connection string");
-            receiver.wal_producer_connstr = wal_producer_connstr.into();
+    mut timeline_updates_receiver: mpsc::UnboundedReceiver<LocalTimelineUpdate>,
+) -> anyhow::Result<()> {
+    let etcd_endpoints = conf.broker_endpoints.clone();
+    ensure!(
+        !etcd_endpoints.is_empty(),
+        "Cannot start wal receiver: etcd endpoints are empty"
+    );
+    let broker_prefix = &conf.broker_etcd_prefix;
+    info!(
+        "Starting wal receiver main thread, etdc endpoints: {}",
+        etcd_endpoints.iter().map(Url::to_string).join(", ")
+    );
+
+    let runtime = tokio::runtime::Builder::new_multi_thread()
+        .thread_name("wal-receiver-runtime-thread")
+        .worker_threads(40)
+        .enable_all()
+        .on_thread_start(|| IS_WAL_RECEIVER.with(|c| c.set(true)))
+        .build()
+        .context("Failed to create storage sync runtime")?;
+    let etcd_client = runtime
+        .block_on(etcd_broker::Client::connect(etcd_endpoints, None))
+        .context("Failed to connect to etcd")?;
+
+    thread_mgr::spawn(
+        ThreadKind::WalReceiverManager,
+        None,
+        None,
+        "WAL receiver manager main thread",
+        true,
+        move || {
+            runtime.block_on(async move {
+                let mut local_timeline_wal_receivers = HashMap::new();
+                loop {
+                    select! {
+                        _ = thread_mgr::shutdown_watcher() => {
+                            info!("Shutdown signal received");
+                            shutdown_all_wal_connections(&mut local_timeline_wal_receivers).await;
+                            break;
+                        },
+                        _ = wal_receiver_main_thread_loop_step(
+                            broker_prefix,
+                            &etcd_client,
+                            &mut timeline_updates_receiver,
+                            &mut local_timeline_wal_receivers,
+                        ) => {},
+                    }
+                }
+            }.instrument(info_span!("wal_receiver_main")));
```

Comment:
BTW, this span in inherited by all subsequent tasks, which is a bit confusing (they are not main). Do you know a way to clear spans for them?